### ISSUE TITLE
Port forwarding not working locally

### ISSUE THREAD

**Issue by jesperpedersen** [containers/libpod#1133]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**Description**

I have an Openshift PostgreSQL image here: https://github.com/jesperpedersen/openshift-pgsql

Once booted the pod doesn't allow access, but it works from within the pod (step 5).

I did try https://github.com/projectatomic/libpod/issues/661#issuecomment-393632515 but that didn't work, hence a new issue.

The `env.txt` file contains
```
PG_DATABASE=mydb
PG_USER_NAME=myuser
PG_USER_PASSWORD=mypass
PG_REPLICATION_NAME=repl
PG_REPLICATION_PASSWORD=replpass
PG_NETWORK_MASK=all
```

**Steps to reproduce the issue:**

1. `systemctl start crio`
2. `buildah bud -t openshift-pgsql10-master-centos7 .`
3. `podman run --env-file env.txt -d -p 5432:5432 openshift-pgsql10-master-centos7:latest`
4. `podman ps`
5. `podman exec -t 342befd98071 /bin/bash`

**Describe the results you received:**

The command `psql -h localhost -p 5432 -U myuser mydb` hangs even though `podman port 342befd98071` gives

```
5432/udp -> 0.0.0.0:5432
5432/tcp -> 0.0.0.0:5432
```

**Describe the results you expected:**

A prompt to the database.

**Additional information you deem important (e.g. issue happens only occasionally):**

RPMs:
```
podman-0.7.3-2.git0791210.fc28.x86_64
containernetworking-cni-0.7.1-1.fc28.x86_64
cri-o-1.11.1-1.git1759204.fc28.x86_64
buildah-1.2-2.gitbe87762.fc28.x86_64
skopeo-0.1.31-5.git0144aa8.fc28.x86_64
```

**Output of `podman version`:**

```
Version:       0.7.3
Go Version:    go1.10.3
OS/Arch:       linux/amd64
```

**Output of `podman info`:**

```
host:
  MemFree: 2993639424
  MemTotal: 16722796544
  SwapFree: 8010067968
  SwapTotal: 8010067968
  arch: amd64
  cpus: 8
  hostname: localhost.localdomain
  kernel: 4.17.6-200.fc28.x86_64
  os: linux
  uptime: 2h 35m 10.15s (Approximately 0.08 days)
insecure registries:
  registries: []
registries:
  registries:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ContainerStore:
    number: 13
  GraphDriverName: overlay
  GraphOptions:
  - overlay.override_kernel_check=true
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
  ImageStore:
    number: 5
  RunRoot: /var/run/containers/storage
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

F28

**Comment by mheon** [comment]
Probably a duplicate of #1119 - though it seems like you might not be able to hit the ports from localhost?

**Comment by mheon** [comment]
Pretty clear we are having significant issues with CNI right now though.

@jesperpedersen Can you try with CRI-O and see if it fails there as well? I want to verify this is definitely a CNI issue.

**Comment by jesperpedersen** [comment]
Yes, this is from localhost.

**Comment by jesperpedersen** [comment]
@mheon Do you have a pointer on how I test with CRI-O ?

**Comment by mheon** [comment]
@jesperpedersen I tried to construct a tutorial, but it's proving to be more trouble than it's worth - I'll follow up with the CRI-O maintainers and see if they're affected

**Comment by mheon** [comment]
@jesperpedersen Can you do a `dnf info containernetworking-cni`? Want to verify if we're seeing the problem on multiple versions

**Comment by jesperpedersen** [comment]
```
Installed Packages
Name         : containernetworking-cni
Version      : 0.7.1
Release      : 1.fc28
Arch         : x86_64
Size         : 35 M
Source       : containernetworking-cni-0.7.1-1.fc28.src.rpm
Repo         : @System
From repo    : @commandline
Summary      : Libraries for writing CNI plugin
URL          : https://github.com/containernetworking/plugins
License      : ASL 2.0
Description  : The CNI (Container Network Interface) project consists of a specification
             : and libraries for writing plugins to configure network interfaces in Linux
             : containers, along with a number of supported plugins. CNI concerns itself
             : only with network connectivity of containers and removing allocated resources
             : when the container is deleted.
```

**Comment by mheon** [comment]
Alright, we're on the same version, then - I think the recent upgrade from 0.6.0 to 0.7.1 broke us. Thanks!

I'm filing a bug against the containernetworking-cni package.

**Comment by jesperpedersen** [comment]
Same issue with
```
Installed Packages
Name         : containernetworking-cni
Version      : 0.6.0
Release      : 4.fc28
Arch         : x86_64
Size         : 33 M
Source       : containernetworking-cni-0.6.0-4.fc28.src.rpm
Repo         : @System
From repo    : @commandline
Summary      : Libraries for writing CNI plugin
URL          : https://github.com/containernetworking/plugins
License      : ASL 2.0
Description  : The CNI (Container Network Interface) project consists of a specification
             : and libraries for writing plugins to configure network interfaces in Linux
             : containers, along with a number of supported plugins. CNI concerns itself
             : only with network connectivity of containers and removing allocated resources
             : when the container is deleted.
```

**Comment by TomSweeneyRedHat** [comment]
FWIW, just tested #1119 against containernetworking-cni 0.6.0 and the latest Podman and it worked without an issue.

**Comment by TomSweeneyRedHat** [comment]
oops, hit the close button by mistake.

**Comment by mheon** [comment]
@TomSweeneyRedHat I managed to reproduce on F27, but that might have been a fluke.

**Comment by mheon** [comment]
Talked with the CNI portmap maintainer, seems like it's a conflict with firewalld.

**Comment by jesperpedersen** [comment]
Doing `systemctl stop firewalld` doesn't help

**Comment by mheon** [comment]
@jesperpedersen I had to disable firewalld and restart the machine to actually have the conflict fixed - can you try that?

**Comment by jesperpedersen** [comment]
Nope, that didn't work either with both containernetworking-cni-0.6.0 and 0.7.1

strace shows:
```
socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 3
setsockopt(3, SOL_TCP, TCP_NODELAY, [1], 4) = 0
fcntl(3, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(3, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
fcntl(3, F_SETFD, FD_CLOEXEC)           = 0
setsockopt(3, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0
connect(3, {sa_family=AF_INET, sin_port=htons(5432), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress)
poll([{fd=3, events=POLLOUT|POLLERR}], 1, -1) = 1 ([{fd=3, revents=POLLOUT}])
getsockopt(3, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
getsockname(3, {sa_family=AF_INET, sin_port=htons(45682), sin_addr=inet_addr("127.0.0.1")}, [128->16]) = 0
poll([{fd=3, events=POLLOUT|POLLERR}], 1, -1) = 1 ([{fd=3, revents=POLLOUT}])
sendto(3, "\0\0\0\10\4\322\26/", 8, MSG_NOSIGNAL, NULL, 0) = 8
poll([{fd=3, events=POLLIN|POLLERR}], 1, -1
```

**Comment by mheon** [comment]
Hi, sorry for the delay here - finally got around to testing/reproducing locally.

I can successfully telnet to 5432 from localhost, so the port is being forwarded.

However, there is definitely a hang in the `psql` process.

**Comment by mheon** [comment]
As such, it looks like port forwarding is working, but something about the traffic is not what `psql` expects, so it's hanging trying to connect.

**Comment by mheon** [comment]
I'm not 100% sure if this is a Podman bug or a `psql` bug, given that traffic seems to be flowing between host and container normally.

**Comment by jesperpedersen** [comment]
The `psql -h localhost -p 5432 -U myuser mydb` works from within the image, so it has to do something with the network bridge.

If that is all external to podman, then we can close here

**Comment by jesperpedersen** [comment]
Adding `--network host` to `podman run` makes the command work

**Comment by mheon** [comment]
I am optimistic that this may have been fixed by #1431 but have not have a chance to test yet

**Comment by jesperpedersen** [comment]
Sorry for the delay - it works with
```
podman-0.10.1.3-1.gitdb08685.fc28.x86_64
containernetworking-plugins-0.7.3-2.fc28.x86_64
cri-o-1.11.2-1.git3eac3b2.fc28.x86_64
buildah-1.4-3.git608fa84.fc28.x86_64
skopeo-0.1.31-13.dev.gite3034e1.fc28.x86_64
```
So, closing.

Thanks !
